turbopuffer/query.py

```diff
@@ -1,40 +1,46 @@
 from dataclasses import dataclass
 import sys
-from typing import Optional, List, Tuple, Union, Dict
-from typing import Iterable
-from typing_extensions import Literal
+from typing import Optional, List, Tuple, Union, Dict, Literal
 
-# Refer to turbopuffer docs for valid operator names
-FilterOperator = str
+FilterOperator = str # https://turbopuffer.com/docs/query
 
 FilterValue = Union[str, int, List[str], List[int]]
 FilterCondition = Tuple[str, FilterOperator, FilterValue]
 
 LegacyFilterCondition = Tuple[FilterOperator, FilterValue]
 LegacyFilterDict = Dict[str, List[LegacyFilterCondition]]
 
-Filters = Union[Tuple[str, List["Filters"]], FilterCondition, LegacyFilterDict]
+Filters = Union[
+    Union[List[Union[str, List]], Tuple[str, List['Filters']]], 
+    FilterCondition,
+    LegacyFilterDict
+]
 
-TextQueryOp = Union[Literal['BM25']]
-TextAggQueryOp = Union[Literal['Sum']]
+# We might type these into literals in the future, similar to FilterOperator
+OrderOperator = str # https://turbopuffer.com/docs/query
+RankByOperator = str # https://turbopuffer.com/docs/query
+SearchOperator = str # https://turbopuffer.com/docs/query
 
-RankInputTextQuery = Union[
-    Tuple[str, TextQueryOp, str],
-    Tuple[TextAggQueryOp, Iterable['RankInputTextQuery']],
-]
+RankByTupleSearch = Tuple[str, SearchOperator, str]
+RankByTupleOrder = Tuple[str, OrderOperator]
+RankByTupleSum   = Tuple[RankByOperator, List[RankByTupleSearch]]
 
-AttributeOrdering = Union[Literal['asc'], Literal['desc']]
+RankByListSearch  = List[Union[str, SearchOperator]]                # e.g. ["text", "BM25", "foo"]
+RankByListOrder   = List[Union[str, OrderOperator]]                 # e.g. ["id", "asc"]
+RankByListSum     = List[Union[RankByOperator, List[RankByTupleSearch]]]  # e.g. ["Sum", [["text", "BM25", "foo"]]]
+RankByListSumList = List[Union[RankByOperator, List[RankByListSearch]]]  # e.g. ["Sum", [["text", "BM25", "foo"]]]
```

Comment:
Do we need these list variants if we have the tuple variants? I feel like the tuple variants are inherently stricter, i.e. technically `["text", "BM25", "hi", "blah", "foo"]` would be fine here since it would match _all_ of the list variants

Whereas using the tuple variants means we'll error if length too long / short

No strong pref. here

Reply:
@morgangallant I'd prefer tuples too, but we use lists all over the place including in examples... so I think we have to do it this way. PITA to type out though.